endpoint: remove policyCalculated field from endpoint

ianvernon opened the pull request:
This field no longer serves much of a purpose, for the following reason:

* `updateNetworkPolicy` is always called after endpoint policy has been
regenerated.
* we calculate policy only in `regenerateBPF`, so we are guaranteed to recompile
the endpoint program if necessary.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5426)
<!-- Reviewable:end -->

Commit:
endpoint: remove policyCalculated field from endpoint

This field no longer serves much of a purpose, for the following reason:

* `updateNetworkPolicy` is always called after endpoint policy has been
regenerated.
* we calculate policy only in `regenerateBPF`, so we are guaranteed to recompile
the endpoint program if necessary.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -416,14 +416,16 @@ func (e *Endpoint) regenerateL3Policy(repo *policy.Repository) (bool, error) {
 	// to get their initial configuration, which is required for them to ACK
 	// the NetworkPolicies.
-	if !e.policyCalculated || e.SecurityIdentity == nil {
+	if e.SecurityIdentity == nil {
```
Perhaps we can check `nextPolicyRevision` for the endpoint here.

raybejjani replied:
Since this function is called after policy is regenerated isn't also true that the endpoint must have an identity? or do we handle `e.SecurityIdentity == nil` as identity Init?

ianvernon replied:
`regeneratePolicy` exits early if the endpoint's `SecurityIdentity` field is nil. I'm not 100 % about whether we can be assured that the Endpoint's security identity is nil or non-nil because identity is assigned via a controller on a per-endpoint basis.

raybejjani commented on pkg/endpoint/policy.go:
```diff
@@ -589,15 +591,6 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (isPolicyComp bool, err error)
 	e.computeDesiredPolicyMapState(repo)
 
-	// If we are in this function, then policy has been calculated.
-	if !e.policyCalculated {
```
I think we had some code that did check the revision number but it might have been changed. @rlenglet was fixing some bug around init generation when I saw this, so he might know more. This might be used to force an initial policy update via the forceRegeneration flag.

ianvernon replied:
Yeah, @rlenglet added a wait for the endpoint to be at policy revision 1 in the code - we bumped the initial policy repository revision to 1, while the endpoint's policy revision is initialized at zero. This means that a non-zero value for the endpoint's policy revision indicates that it has been entirely regenerated (i.e. policy, BPF program, PolicyMap, proxy) successfully and has had its revision bumped. Also, I noticed that we don't even check the result of `regeneratePolicy` with respect to whether the policy has changed anymore because we always try and see if the endpoint needs to be regenerated after, so I think I can go ahead and get rid of that return value from it as well since it's somewhat misleading now.

ianvernon replied:
To clarify - the reason we don't need this anymore is because it doesn't affect what is done after we call `regeneratePolicy` - the returned value isn't used anywhere, so we aren't changing any functional behavior anymore.

Commit:
endpoint: check whether endpoint has Identity right before policy is regenerated

Move this to the start of `regeneratePolicy` as no policy can be calculated without an identity.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove boolean return value from `regeneratePolicy`

The result of this is no longer used anywhere in the code base; remove it. We
still log whether any meaningful changes have been applied for the endpoint at
the end of said function, so we are not losing any visibility into policy
calculation here.

Signed-off by: Ian Vernon <ian@cilium.io>
